cachito/workers/scm.py

```diff
@@ -0,0 +1,201 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import logging
+import os
+from abc import ABC, abstractmethod
+import urllib.parse
+import shutil
+import tarfile
+import tempfile
+import subprocess
+
+import requests
+
+from cachito.errors import ConfigError, CachitoError
+
+
+log = logging.getLogger(__name__)
+
+
+# TODO: Write tests
+class SCM(ABC):
+    """The base class for interacting with source control."""
+
+    def __init__(self, url, ref):
+        """
+        Initialize the SCM class.
+
+        :param str url: the source control URL to the repository to fetch
+        :param str ref: the source control reference
+        """
+        super().__init__()
+        self.url = url
+        self.ref = ref
+        self._archives_dir = None
+        self._archive_path = None
+        self._repo_name = None
+        # Validate that the base directory can be retrieved
+        self.archives_dir
+
+    @property
+    def archive_path(self):
+        """
+        Get the path to where the archive for a particular SCM reference should be.
+
+        :return: the path to the archive
+        :rtype: str
+        """
+        if not self._archive_path:
+            directory = os.path.join(self.archives_dir, *self.repo_name.split('/'))
+            # Create the directories if they don't exist
+            os.makedirs(directory, exist_ok=True)
+            self._archive_path = os.path.join(directory, f'{self.ref}.tar.gz')
+
+        return self._archive_path
+
+    @property
+    def archives_dir(self):
+        """
+        Get the absolute path of the archives directory from the Celery configuration.
+
+        :returns: the absolute path of the archives directory
+        :rtype: str
+        :raise ConfigError: if "cachito_archives_dir" is not configured in Celery
+        """
+        if not self._archives_dir:
+            # Import this here to avoid a circular import
+            from cachito.workers.tasks import app as celery_app
+
+            if not celery_app.conf.get('cachito_archives_dir'):
```

Comment:
Can we make this a required attribute in the config?  I'd rather see an error during app start time then something that happens later on. IOW, let's move this to `validate_celery_config`.

Reply:
Is it safe to assume that all of the celery workers will need access to the archives directory? If so, I can make it mandatory.